File: app/src/main/java/com/nbennettsoftware/android/npad/SettingsFragment.java
package com.nbennettsoftware.android.npad;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;

import java.util.List;

public class SettingsFragment extends PreferenceFragment {

    private Utils utils;
    private int PICK_WALLPAPER_INTENT_ID = 0;
    public OnWallpaperChangedListener onWallpaperChangedListener = null;
    public OnShadeChangedListener onShadeChangedListener = null;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        utils = new Utils(this.getActivity());

        // Load the preferences from an XML resource
        addPreferencesFromResource(R.xml.preferences);
        bindPreferenceSummaryToValue(findPreference(getString(R.string.pref_key_font_size)));
        bindPreferenceSummaryToValue(findPreference(getString(R.string.pref_key_shade_intensity)));

        //We let bindPrefernceSummaryToValue set are summary for us.
        findPreference(getString(R.string.pref_key_shade_intensity)).setOnPreferenceChangeListener(new OnShadeIntensityChangeListener());

        findPreference(getString(R.string.pref_key_pick_wallpaper)).setOnPreferenceClickListener(new OnPickWallpaperClickListener());
        findPreference(getString(R.string.pref_key_clear_wallpaper)).setOnPreferenceClickListener(new OnClearWallpaperClickListener());
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(requestCode == PICK_WALLPAPER_INTENT_ID && resultCode == Activity.RESULT_OK) {
            Uri uri = data.getData();
            if(uri==null) { return; }
            utils.saveWallpaper(uri);
            if(onWallpaperChangedListener != null) {
                onWallpaperChangedListener.OnWallpaperChanged();
            }
        }
    }

    interface OnWallpaperChangedListener {
        void OnWallpaperChanged();
    }

    interface OnShadeChangedListener {
        void OnShadeChanged(String shadeIntensity);
    }

    void setOnWallpaperChangedListener(OnWallpaperChangedListener onWallpaperChangedListener) {
        this.onWallpaperChangedListener = onWallpaperChangedListener;
    }

    void setOnShadeChangedListener(OnShadeChangedListener onShadeChangedListener) {
        this.onShadeChangedListener = onShadeChangedListener;
    }

    private class OnPickWallpaperClickListener implements Preference.OnPreferenceClickListener {
        @Override
        public boolean onPreferenceClick(Preference preference) {
            final String IMAGE_MIME_TYPE="image/*";
            Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
            intent.setType(IMAGE_MIME_TYPE);
            intent.addCategory(Intent.CATEGORY_OPENABLE);

            Intent chooser = Intent.createChooser(intent, "Select Image");

            if (chooser.resolveActivity(getActivity().getPackageManager()) != null) {
                startActivityForResult(chooser, PICK_WALLPAPER_INTENT_ID);
            } else {
                utils.toast("No apps installed.");
            }
            return true;
        }
    }

    private class OnClearWallpaperClickListener implements Preference.OnPreferenceClickListener {
        @Override
        public boolean onPreferenceClick(Preference preference) {
            utils.clearWallpaper();
            if(onWallpaperChangedListener != null) {
                onWallpaperChangedListener.OnWallpaperChanged();
            }
            return true;
        }
    }

    private class OnShadeIntensityChangeListener implements Preference.OnPreferenceChangeListener {
        @Override
        public boolean onPreferenceChange(Preference preference, Object shadeIntensityObj) {
            //Update summery
            sBindPreferenceSummaryToValueListener.onPreferenceChange(preference, shadeIntensityObj);
            if(onShadeChangedListener != null) {
                onShadeChangedListener.OnShadeChanged(shadeIntensityObj.toString());
            }
            return true;
        }
    }

    private static Preference.OnPreferenceChangeListener sBindPreferenceSummaryToValueListener = new Preference.OnPreferenceChangeListener() {
        @Override
        public boolean onPreferenceChange(Preference preference, Object value) {
            String stringValue = value.toString();

            if (preference instanceof ListPreference) {
                ListPreference listPreference = (ListPreference) preference;
                int index = listPreference.findIndexOfValue(stringValue);
                preference.setSummary(
                        index >= 0
                                ? listPreference.getEntries()[index]
                                : null);
            } else {
                preference.setSummary(stringValue);
            }
            return true;
        }
    };

    private static void bindPreferenceSummaryToValue(Preference preference) {
        preference.setOnPreferenceChangeListener(sBindPreferenceSummaryToValueListener);

        sBindPreferenceSummaryToValueListener.onPreferenceChange(preference,
                PreferenceManager
                        .getDefaultSharedPreferences(preference.getContext())
                        .getString(preference.getKey(), ""));
    }
}

File: app/src/main/java/com/nbennettsoftware/android/npad/SettingsFragment.kt
package com.nbennettsoftware.android.npad

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.preference.*

import com.nbennettsoftware.android.npad.storage.WallpaperManager

const val PICK_WALLPAPER_INTENT_ID = 0

class SettingsFragment : PreferenceFragmentCompat() {

    private lateinit var wallpaperManager: WallpaperManager
    var onWallpaperChangedListener: OnWallpaperChangedListener? = null
    var onScalingChangedListener: OnScalingChangedListener? = null
    var onShadeChangedListener: OnShadeChangedListener? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        wallpaperManager = WallpaperManager(activity as Activity)
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences, rootKey)

        bindPreferenceSummaryToValue(findPreference(getString(R.string.pref_key_font_size)))
        bindPreferenceSummaryToValue(findPreference(getString(R.string.pref_key_dimmer_intensity)))
        bindPreferenceSummaryToValue(findPreference(getString(R.string.pref_key_scaling)))

        //We let bindPreferenceSummaryToValue set are summary for us.
        findPreference(getString(R.string.pref_key_scaling)).onPreferenceChangeListener = OnScalingChangeListener()
        findPreference(getString(R.string.pref_key_dimmer_intensity)).onPreferenceChangeListener = OnShadeIntensityChangeListener()

        findPreference(getString(R.string.pref_key_pick_wallpaper)).onPreferenceClickListener = OnPickWallpaperClickListener()
        findPreference(getString(R.string.pref_key_clear_wallpaper)).onPreferenceClickListener = OnClearWallpaperClickListener()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == PICK_WALLPAPER_INTENT_ID && resultCode == Activity.RESULT_OK) {
            val uri = data.data ?: return
            try {
                wallpaperManager.replaceInternalizeWallpaper(uri)
            } catch (e: WallpaperManager.ReplaceInternalWallpaperException) {
                e.printStackTrace()
                Toast.makeText(activity, "Cannot replace internal wallpaper", Toast.LENGTH_LONG).show()
            }

            if (onWallpaperChangedListener != null) {
                onWallpaperChangedListener!!.onWallpaperChanged()
            }
        }
    }

    interface OnWallpaperChangedListener {
        fun onWallpaperChanged()
    }

    interface OnScalingChangedListener {
        fun onScalingChanged(scaling: String)
    }

    interface OnShadeChangedListener {
        fun onDimmerChanged(shadeIntensity: String)
    }

    internal fun setOnWallpaperChangedListener(onWallpaperChangedListener: OnWallpaperChangedListener) {
        this.onWallpaperChangedListener = onWallpaperChangedListener
    }

    internal fun setOnScalingChangedListener(onScalingChangedListener: OnScalingChangedListener) {
        this.onScalingChangedListener = onScalingChangedListener
    }

    internal fun setOnShadeChangedListener(onShadeChangedListener: OnShadeChangedListener) {
        this.onShadeChangedListener = onShadeChangedListener
    }

    private inner class OnPickWallpaperClickListener : Preference.OnPreferenceClickListener {
        override fun onPreferenceClick(preference: Preference): Boolean {
            val imageMimeType = "image/*"
            val intent = Intent(Intent.ACTION_GET_CONTENT)
            intent.type = imageMimeType
            intent.addCategory(Intent.CATEGORY_OPENABLE)

            if (intent.resolveActivity(activity!!.packageManager) != null) {
                startActivityForResult(intent, PICK_WALLPAPER_INTENT_ID)
            } else {
                Toast.makeText(activity, "No apps installed.", Toast.LENGTH_SHORT).show()
            }
            return true
        }
    }

    private inner class OnClearWallpaperClickListener : Preference.OnPreferenceClickListener {
        override fun onPreferenceClick(preference: Preference): Boolean {
            val preferences = PreferenceManager.getDefaultSharedPreferences(activity)
            val drawDefaultBackground_key = activity!!.resources.getString(R.string.pref_key_draw_default_background)
            val drawDefaultBackground_default = activity!!.resources.getBoolean(R.bool.pref_default_draw_default_background)

            val drawDefaultBackground = preferences.getBoolean(drawDefaultBackground_key, drawDefaultBackground_default)

            if (wallpaperManager.hasInternalizedWallpaper()) {
                wallpaperManager.deleteInternalizedWallpaper()
            } else {
                preferences.edit().putBoolean(drawDefaultBackground_key, !drawDefaultBackground).apply()
            }
            if (onWallpaperChangedListener != null) {
                onWallpaperChangedListener!!.onWallpaperChanged()
            }
            return true
        }
    }

    private inner class OnScalingChangeListener : Preference.OnPreferenceChangeListener {
        override fun onPreferenceChange(preference: Preference, scalingObj: Any): Boolean {
            //Update summery
            sBindPreferenceSummaryToValueListener.onPreferenceChange(preference, scalingObj)
            if (onScalingChangedListener != null) {
                onScalingChangedListener!!.onScalingChanged(scalingObj.toString())
            }
            return true
        }
    }

    private inner class OnShadeIntensityChangeListener : Preference.OnPreferenceChangeListener {
        override fun onPreferenceChange(preference: Preference, shadeIntensityObj: Any): Boolean {
            //Update summery
            sBindPreferenceSummaryToValueListener.onPreferenceChange(preference, shadeIntensityObj)
            if (onShadeChangedListener != null) {
                onShadeChangedListener!!.onDimmerChanged(shadeIntensityObj.toString())
            }
            return true
        }
    }

    companion object {

        private val sBindPreferenceSummaryToValueListener = Preference.OnPreferenceChangeListener { preference, value ->
            val stringValue = value.toString()

            if (preference is ListPreference) {
                val index = preference.findIndexOfValue(stringValue)
                preference.setSummary(
                        if (index >= 0)
                            preference.entries[index]
                        else
                            null)
            } else {
                preference.summary = stringValue
            }
            true
        }

        private fun bindPreferenceSummaryToValue(preference: Preference) {
            preference.onPreferenceChangeListener = sBindPreferenceSummaryToValueListener

            sBindPreferenceSummaryToValueListener.onPreferenceChange(preference,
                    PreferenceManager
                            .getDefaultSharedPreferences(preference.context)
                            .getString(preference.key, ""))
        }
    }
}
